aptos-move/framework/aptos-stdlib/sources/data_structures/smart_table.move

```diff
@@ -180,6 +185,82 @@ module aptos_std::smart_table {
         res
     }
 
+    /// Get all keys in a smart table.
+    ///
+    /// For a large enough smart table this function will fail due to execution gas limits, and
+    /// `keys_paginated` should be used instead.
+    public fun keys<K: store + copy + drop, V: store + copy>(
+        table_ref: &SmartTable<K, V>
+    ): vector<K> {
+        let (keys, _, _) = keys_paginated(table_ref, 0, 0, length(table_ref));
+        keys
+    }
+
+    /// Get keys from a smart table, paginated.
+    ///
+    /// This function can be used to paginate all keys in a large smart table outside of runtime,
+    /// e.g. through chained view function calls. The maximum `num_keys_to_get` before hitting gas
+    /// limits depends on the data types in the smart table.
+    ///
+    /// When starting pagination, pass `starting_bucket_index` = `starting_vector_index` = 0.
+    ///
+    /// The function will then return a vector of keys, an optional bucket index, and an optional
+    /// vector index. The unpacked return indices can then be used as inputs to another pagination
+    /// call, which will return a vector of more keys. This process can be repeated until the
+    /// returned bucket index and vector index value options are both none, which means that
+    /// pagination is complete. For an example, see `test_keys()`.
+    public fun keys_paginated<K: store + copy + drop, V: store + copy>(
+        table_ref: &SmartTable<K, V>,
+        starting_bucket_index: u64,
+        starting_vector_index: u64,
+        num_keys_to_get: u64,
+    ): (
+        vector<K>,
+        Option<u64>,
+        Option<u64>,
+    ) {
+        let num_buckets = table_ref.num_buckets;
+        let buckets_ref = &table_ref.buckets;
+        assert!(starting_bucket_index < num_buckets, EINVALID_BUCKET_INDEX);
+        let bucket_ref = table_with_length::borrow(buckets_ref, starting_bucket_index);
+        let bucket_length = vector::length(bucket_ref);
+        assert!(
+            // In the general case, starting vector index should never be equal to bucket length
+            // because then iteration will attempt to borrow a vector element that is out of bounds.
+            // However starting vector index can be equal to bucket length in the special case of
+            // starting iteration at the beginning of an empty bucket since buckets are never
+            // destroyed, only emptied.
+            starting_vector_index < bucket_length || starting_vector_index == 0,
```

Comment:
starting_vector_index <= bucket_length?

Reply:
@lightmark  No, as specified in the comment: 

```
In the general case, starting vector index should never be equal to bucket length because then iteration will attempt to borrow a vector element that is out of bounds.

However starting vector index can be equal to bucket length in the special case of starting iteration at the beginning of an empty bucket since buckets are never destroyed, only emptied
```

If this were to use `starting_vector_index <= bucket_length` as you suggest then it would be possible to attempt borrowing out of bounds on a vector

Reply:
`starting_vector_index..bucket_length` will not loop so borrow would not happen in this way. Am I wrong?

Reply:
You are correct that the loop will not borrow

But that is a silent form of feedback for non-well-formed inputs, which does not allow a user to know that their pagination has failed.

E.g. if the bucket length is 5 and I pass 5 then I want an error so that I know that I have fumbled the indexing

Reply:
starting_vector_index = 0 actually is already a non-well-formed input when bucket_length = 0.
I think allow starting_vector_index == bucket_length is similar to `iterator.end()` in c++, which is not a valid iterator to get value but as a sentinel value.

Reply:
> starting_vector_index = 0 actually is already a non-well-formed input when bucket_length = 0.

This is actually a typical operation during valid pagination: if `num_keys_to_get` is reached just at the end of the bucket, the algorithm indicates to start the next pagination call at the beginning of the next bucket:

https://github.com/alnoki/aptos-core/blob/34953c14d92ef7b1ad6c377a15542bbef0a97724/aptos-move/framework/aptos-stdlib/sources/data_structures/smart_table.move#L245-L250

Then if the next bucket is empty, you'll start the next pagination pass by passing `starting_vector_index = 0` for an empty bucket.

The alternative is, if `num_keys_to_get` is reached, to keep looping until finding the beginning of the next non-empty bucket, but this is a waste of gas if the user has no intention of paginating again

Reply:
Maybe I am paranoid. I mean `starting_vector_index <= bucket_length` seems better because if you allow (0, 0) but not allow (5, 5) it is not consistent from api perspective.
Even you allow (5, 5), the behavior is the same but the valid argument rule is more clear.